cudamapper/include/claraparabricks/genomeworks/cudamapper/index.hpp

```diff
@@ -88,71 +88,92 @@ class Index
     virtual read_id_t largest_read_id() const = 0;
 
     /// \brief returns length of the longest read in this index
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return length of the longest read in this index
     virtual position_in_read_t number_of_basepairs_in_longest_read() const = 0;
 
     /// \brief Return the maximum kmer length allowable
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return Return the maximum kmer length allowable
     static uint64_t maximum_kmer_size()
     {
         return sizeof(representation_t) * CHAR_BIT / 2;
     }
 
+    /// \brief checks if index is ready to be used on this device, index might not be ready if its creation or copy from host is asynchronous
+    /// \return true if the index is ready to be used
+    virtual bool is_ready() const = 0;
+
+    /// \brief if is_ready() is true returns immediately, blocks until it becomes ready otherwise
+    virtual void wait_to_be_ready() = 0;
+
     /// \brief generates a mapping of (k,w)-kmer-representation to all of its occurrences for one or more sequences
+    /// Generation is done asynchronously and one should call wait_to_be_ready() before using the object
     /// \param allocator The device memory allocator to use for temporary buffer allocations
     /// \param parser parser for the whole input file (part that goes into this index is determined by first_read_id and past_the_last_read_id)
     /// \param descriptor object holding the first read, size of the read, and hash value of the index being created
     /// \param kmer_size k - the kmer length
     /// \param window_size w - the length of the sliding window used to find sketch elements  (i.e. the number of adjacent kmers in a window, adjacent = shifted by one basepair)
     /// \param hash_representations if true, hash kmer representations
     /// \param filtering_parameter filter out all representations for which number_of_sketch_elements_with_that_representation/total_skech_elements >= filtering_parameter, filtering_parameter == 1.0 disables filtering
-    /// \param cuda_stream CUDA stream on which the work is to be done. Device arrays are also associated with this stream and will not be freed at least until all work issued on this stream before calling their destructor is done
+    /// \param cuda_stream_generation CUDA stream on which index is generated. Device arrays are associated with this stream and will not be freed at least until all work issued on this stream before calling their destructor has been done
+    /// \param cuda_stream_copy CUDA stream on which the index is copied to host if needed. Device arrays are associated with this stream and will not be freed at least until all work issued on this stream before calling their destructor has been done
     /// \return instance of Index
     static std::unique_ptr<Index>
     create_index(DefaultDeviceAllocator allocator,
```

Comment:
Also here, would it make sense to rename this to `create_index_async`?
I'm a bit ambiguous, because the function is mostly, but not fully asynchronous.

Reply:
Let's say that for now `_async` means "not all GPU work is done when this function returns"
done